### PR TITLE
Add the C_DEBUGGER preprocessor symbol

### PR DESCRIPTION
**Pull request by naerbnic**
# Description

Added C_DEBUGGER config preprocessor symbol

## Related issues

_Related issues can be listed here (remove the section if not applicable.)_

# Manual testing

Compiled using the `-Denable_debugger=<none|normal|heavy>`, and checked that it behaved as expected

_Provide clear instructions so the reviewers can reproduce and verify your results._

_Include relevant details for your test configuration, operating system, etc._

_Ideally, you would also test your changes with a [sanitizer build](https://github.com/dosbox-staging/dosbox-staging/blob/main/BUILD.md#make-a-sanitizer-build) and confirm no issues were found._

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [ ] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [ ] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -404,14 +404,9 @@ conf_data.set10('C_MT32EMU', get_option('use_mt32emu'))
...
-
-if get_option('enable_debugger') != 'none'
-    conf_data.set10('C_DEBUG', true)
-endif
-
-if get_option('enable_debugger') == 'heavy'
-    conf_data.set10('C_HEAVY_DEBUG', true)
-endif
+conf_data.set10('C_DEBUGGER', get_option('enable_debugger') != 'none')
+conf_data.set10('C_HEAVY_DEBUG', get_option('enable_debugger') == 'heavy')
```
Comment (johnnovak) [inline meson.build:408]: Maybe a consistency change to `C_HEAVY_DEBUGGER`? Now you could think it's a heavy debug logging mode or something.

**Comment by johnnovak** [comment]
Had a quick look, it looks good @naerbnic. The doco Markdown files and the GitHub workflow should also be updated:

image

@kcgen We don't want to merge this until 0.81 is out, right? I guess it would muddy the water and complicate things even if it's just in `main`.

**Comment by kcgen** [comment]
> @kcgen We don't want to merge this until 0.81 final is out, right? 

I think that's the easiest and safest approach. That way `main` almost entirely accumulates the identical fixes that will be cherry-picked over to the 0.81.x branch, without having to "step over" large changes like this.

The only catch is that @naerbnic will need to keep this branch re-based as `main` continues to collect more backport commits.

**Comment by johnnovak** [comment]
> > @kcgen We don't want to merge this until 0.81 final is out, right?
> 
> I think that's the easiest and safest approach. That way `main` almost entirely accumulates the identical fixes that will be cherry-picked over to the 0.81.x branch, without having to "step over" large changes like this.

Agreed, let's do that. I like simple. Simple is good 😎 🤘🏻 

> The only catch is that @naerbnic will need to keep this branch re-based as `main` continues to collect more backport commits.

Please keep rebasing your feature branch @naerbnic until 0.81 final comes out, then we'll merge this to main. I think it's fine to have a single branch for all the debugger enhancements as long as you make small, bisectable commits per logical chunks of work.

Ideally, we can release the final version sometime in mid-late Februrary. Definitely not in Jan.

**Comment by FeralChild64** [comment]
How does one use the debugger? ALT+PAUSE indeed shows the PDCurses debugger window for me, but DOSBox stops reacting on key presses and both windows are not redrawn anymore.

**Comment by naerbnic** [comment]
> Please keep rebasing your feature branch @naerbnic until 0.81 final comes out, then we'll merge this to main. I think it's fine to have a single branch for all the debugger enhancements as long as you make small, bisectable commits per logical chunks of work.

I'm probably going to wait for a bit to continue until after the 0.81 final in general. Then I can do a rebase (which while it will have a bunch of changes, I feel should provide for fairly simple merging for this change) without having to keep it continuously up to date. Tell me if this is a bad plan for this codebase.

> How does one use the debugger? ALT+PAUSE indeed shows the PDCurses debugger window for me, but DOSBox stops reacting on key presses and both windows are not redrawn anymore.

That's odd. At very least typing on the debugger window should change the debugger command line. For what it's worth, I've tried this mainly on Windows. If you're talking about the main DOSBox window, it will freeze until you resume the program in the debugger.

I put together a [very short doc](https://docs.google.com/document/d/1e0vRuDl5dH7u6B4aZRCKcNXjiz_VYjx9wx1cAz76-QI/edit?usp=sharing) about the debugger which at very least shows my experience, and maybe a few tips in addition

File: meson.build (hunk 1 above)
Comment (naerbnic) [reply meson.build:408]: Well, in some cases, it is. I need to do a more thorough examination, but I think at least some of the C_HEAVY_DEBUG-wrapped code is more thorough logging of some DOSBox internals. I'll take a look and either split it in another commit, or change it soon

**Comment by FeralChild64** [comment]
Well, the debugger window is completely frozen for me, it does not react on keypresses at all. I can’t type HELP there. If I start moving windows around, debugger window eventually loses it’s content and is not redrawn.

I’m using Linux almost exclusively - I only switch to Windows if I really have no other choice. I have never managed to run the debugger succesfully.

**Comment by johnnovak** [comment]
> I'm probably going to wait for a bit to continue until after the 0.81 final in general. Then I can do a rebase (which while it will have a bunch of changes, I feel should provide for fairly simple rebasing for this change) without having to keep it continuously up to date. Tell me if this is a bad plan for this codebase.

Yep, that's a good plan. We'll be able to focus more on this after 0.81 is out.



> I’m using Linux almost exclusively - I only switch to Windows if I really have no other choice. I have never managed to run the debugger succesfully.

Maybe it's Windows only. I will try it on macOS... except you don't get a Pause key on MacBooks, so I guess I'll need to hack it a bit 😆

**Comment by johnnovak** [comment]
@naerbnic Thanks for the patience. Now is the time to finalise this, then the team will review it in the coming weeks.

**Comment by johnnovak** [comment]
Closing due to inactivity. Please feel free to raise another PR if you want to finish the work @naerbnic.

**Pull request closed without merging**